sql: panic with unsupported comparison

**Describe the problem**

The latest version of the CockroachDB (dc93fa8ccfc0) crashes when executing the following query:

```sql
SET vectorize = off;
CREATE TABLE v0(c1 TIME PRIMARY KEY, c2 INT);
INSERT INTO v0 VALUES('0:0:0'::TIME, 0);
SELECT * FROM v0 WHERE(c1, c2)=(SELECT 0.0, 0);
```

**To Reproduce**

Here is the detail steps to reproduce the bug.

1. In system Ubuntu 20.04, download the CockroachDB source code from the github source.
2. Use the latest version of the CockroachDB code (tested version: dc93fa8ccfc0)
3. Directly `make install` in the root repository folder. 
4. Run `./cockroach demo`, and then paste the PoC query to the cockroach cli environment.
5. Observe the crashing and log the crashing information. 

**Expected behavior**
The DBMS should throw an error such as the following: `pq: unsupported comparison: time to decimal`. 

**Additional data / screenshots**

Here is the crashing stack trace: 

```
*
* ERROR: a SQL panic has occurred while executing the following statement:
* SELECT * FROM v0 WHERE (c1, c2) = (SELECT 0.0, 0)
*
*
* ERROR: a panic has occurred!
* flow: unexpected 10240 leftover bytes
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:392
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:445
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:420
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:581
*   | github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:442
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func2
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:791
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:890
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTime).Compare
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:2296
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalCompareTupleOp
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/binary_op.go:201
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CompareTupleOp).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_op_generated.go:301
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalComparisonExpr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:267
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ComparisonExpr).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_expr_generated.go:121
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.RunFilter
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:229
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*ExprHelper).EvalFilter
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:218
*   | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*filtererProcessor).Next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/filterer.go:98
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:186
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:733
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:483
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:815
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1768
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1520
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1624
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1245
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:737
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2513
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1955
* Wraps: (2) while executing: SELECT * FROM _ WHERE (_, _) = (SELECT _, _)
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:392
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:445
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:420
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:581
*   | github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:442
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func2
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:791
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:890
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTime).Compare
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:2296
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalCompareTupleOp
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/binary_op.go:201
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CompareTupleOp).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_op_generated.go:301
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalComparisonExpr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:267
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ComparisonExpr).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_expr_generated.go:121
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.RunFilter
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:229
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*ExprHelper).EvalFilter
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:218
*   | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*filtererProcessor).Next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/filterer.go:98
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:186
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:733
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:483
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:815
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1768
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1520
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1624
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1245
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:737
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2513
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1955
* Wraps: (4) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:390
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:445
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:420
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:581
*   | github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:442
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func2
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:791
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:890
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTime).Compare
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:2296
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalCompareTupleOp
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/binary_op.go:201
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CompareTupleOp).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_op_generated.go:301
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalComparisonExpr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:267
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ComparisonExpr).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_expr_generated.go:121
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.RunFilter
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:229
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*ExprHelper).EvalFilter
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:218
*   | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*filtererProcessor).Next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/filterer.go:98
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:186
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:733
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:483
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:815
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1768
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1520
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1624
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1245
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:737
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2513
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1955
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1960
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1878
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:834
* Wraps: (5) flow: unexpected 10240 leftover bytes
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *withstack.withStack (5) *errutil.leafError
*
panic: unsupported comparison: time to decimal
	panic: flow: unexpected 10240 leftover bytes [recovered]
	panic: flow: unexpected 10240 leftover bytes

goroutine 5051 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc002ab5000, {0x82aa4d0, 0xc003518700}, {0x5054ca0?, 0xc00595e168?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1080 +0x1b4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x82aa4d0, 0xc003518700}, {0x11dd8d4?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832 +0x4a
panic({0x5054ca0, 0xc00595e168})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x82aa4d0, 0xc00314e5c0}, 0xc00395c000, {0x54e2595?, 0x82e3c00?}, {0xc005777330?, 0xc406b8?, 0xc003492c58?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:392 +0x1c5
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc003529040, {0x82aa4d0, 0xc00314e5c0}, 0x1)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:445 +0x21a
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:420
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0xc0042eca00, {0x82aa4d0, 0xc00314e5c0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:581 +0x27e
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0xc00592b260, {0x82aa4d0?, 0xc00314e5c0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:442 +0x2e
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func2()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:791 +0x2a
panic({0x4f5cfe0, 0xc003b11140})
	/usr/local/go/src/runtime/panic.go:890 +0x262
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTime).Compare(0xc002649c00?, {0x82c6118?, 0xc002649c00?}, {0x830cff0?, 0xc00122fd90?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:2296 +0xc8
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalCompareTupleOp(0xc003b11120?, {0xc002649c00?, 0xc005777680?}, 0xc00038b440, {0x830db38?, 0xc00505b1a0?}, {0x830db38?, 0xc0047405d0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/binary_op.go:201 +0x330
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CompareTupleOp).Eval(0x98fd64e180f1d?, {0x82aa578?, 0xc004741140?}, {0x8364c38?, 0xc002649c00?}, {0x830db38?, 0xc00505b1a0?}, {0x830db38?, 0xc0047405d0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_op_generated.go:301 +0x62
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalComparisonExpr(0x10?, {0x82aa578, 0xc004741140}, 0xc002625a40)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:267 +0x2a9
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ComparisonExpr).Eval(0x0?, {0x82aa578?, 0xc004741140?}, {0x833a9d0?, 0xc002649c00?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_expr_generated.go:121 +0x34
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr(...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.RunFilter({0x82aa578?, 0xc004741140?}, {0x82dc608?, 0xc002625a40?}, 0xc0dd5b105218e7cf?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:229 +0x48
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*ExprHelper).EvalFilter(0xc00331d8c0, {0x82aa578?, 0xc004741140?}, {0xc003568000?, 0x36?, 0x0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/expr.go:218 +0x190
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*filtererProcessor).Next(0xc0040a6000)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/filterer.go:98 +0xa6
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run({0x82aa578, 0xc004741140}, {0x82cff00, 0xc0040a6000}, {0x8281a18?, 0xc00507f500?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:186 +0x55
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc0040a6000, {0x82aa4d0?, 0xc00314e5c0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:733 +0x5b
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc0042eca00, {0x82aa4d0, 0xc00314e5c0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:483 +0x254
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000ad7040, {0x82aa578, 0xc004740840}, 0xc001b34000, 0xc005ba6840, 0xc0048d9200, 0xc00670c000, 0xc002ab58b0, 0x0)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:815 +0x984
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x1?, {0x82aa578, 0xc004740840}, 0xc004740840?, 0xc001b34000, 0xc0057783c0?, {{0x82ac5a0?, 0xc0048d8e80?}, 0x0?}, 0xc00670c000)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1768 +0x1f9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc002ab5000?, {0x82aa578, 0xc004740840}, 0xc00046d500?, 0xc002ab5540?, 0xc002ab5540, 0xc00670c000, 0xc003ec67e0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1520 +0x206
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc002ab5000, {0x82aa578, 0xc004740840}, 0xc002ab5540, 0xc004740840?, {0x7fb874fae4b0?, 0xc002cedf10}, 0x0?, 0xc00504f7d8)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1624 +0x4c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc002ab5000, {0x82aa578, 0xc004740120}, 0xc002ab5540, {0x7fb874fae4b0, 0xc002cedf10})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1245 +0xc13
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc002ab5000, {0x82aa578?, 0xc005457ef0?}, {{0x82d3800, 0xc0026254a0}, {0x0, 0x0, 0x0}, {0xc003a42460, 0x2f}, ...}, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:737 +0x2634
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x82aa578?, 0xc005457ef0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130 +0xbc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x82aa578?, {0x82aa578?, 0xc005457ef0?}, {0x82d3800?, 0xc0026254a0?}, 0x0?, 0xaf47700?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2513 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc002ab5000, {0x82aa578, 0xc005457ef0}, {{0x82d3800, 0xc0026254a0}, {0x0, 0x0, 0x0}, {0xc003a42460, 0x2f}, ...}, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129 +0x4fa
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc005779ac0, 0xc002ab5000, 0xc0057797b8, 0xc005779870, 0xc0057798b0, 0xc005779890, 0xc005779880, 0xc0057798a0)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1955 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc002ab5000)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1960 +0x1127
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc002ab5000, {0x82aa4d0, 0xc003518700}, 0x418b5efff0000000?, 0xc002ca1dd8?, 0xc0058db860?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1878 +0x20b
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc00354a320?, {0x82aa4d0?, 0xc003518700?}, {0xc0016ace00?}, 0x3?, 0xc00371fe80?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:834 +0xe6
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:728 +0x40d
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:639 +0x22e
```

**Environment:**
 - CockroachDB version: v23.1.0-alpha
 - Server OS: Ubuntu 20.04 LTS
 - Client app: CockroachDB demo command. (`./cockroach demo`)
 - Detailed CockroachDB version output:

```
Build Tag:        v23.1.0-alpha.00000000-1118-gdc93fa8ccf-dirty
Build Time:       2022/12/10 21:03:15
Distribution:     CCL
Platform:         linux amd64 (x86_64-linux-gnu)
Go Version:       go1.19
C Compiler:       gcc 9.4.0
Build Commit ID:  dc93fa8ccfc0d08afd81a3222426d5608fd7a529
Build Type:       development
```

**Additional context**

The problem is reproducible on all versions from `v22.2 (including the Production Releases)`, `v22.1` and `v21.2`. I did not test the even earlier versions so I do not know when this crashing problem is introduced. 


Jira issue: CRDB-22310

--- a/pkg/col/coldataext/datum_vec.go
+++ b/pkg/col/coldataext/datum_vec.go
@@ -55,7 +55,15 @@ func newDatumVec(t *types.T, n int, evalCtx *eval.Context) coldata.DatumVec {
 // Note that the method is named differently from "Compare" so that we do not
 // overload tree.Datum.Compare method.
 func CompareDatum(d, dVec, other interface{}) int {
-	return d.(tree.Datum).Compare(dVec.(*datumVec).evalCtx, convertToDatum(other))
+	// Note that it's not strictly necessary to use CompareError here instead of
+	// just Compare since pkg/sql/sem/eval is in the allow-list of paths that
+	// colexecerror catches panics from. Still, it seems nicer to be explicit
+	// about this.
+	res, err := d.(tree.Datum).CompareError(dVec.(*datumVec).evalCtx, convertToDatum(other))
+	if err != nil {
+		colexecerror.InternalError(err)
+	}
+	return res
 }
 
 // Hash returns the hash of the datum as a byte slice.

--- a/pkg/sql/logictest/testdata/logic_test/tuple_local
+++ b/pkg/sql/logictest/testdata/logic_test/tuple_local
@@ -1,0 +1,13 @@
+# LogicTest: !fakedist !fakedist-vec-off !fakedist-disk !3node-tenant
+
+# At the moment, the query below results in an internal error when executed in
+# a distributed fashion. We should fix that and merge this file into 'tuple'
+# (tracked in #94970).
+
+# Regression test for not using CompareError when dealing with tuples (#93396).
+statement ok
+CREATE TABLE t93396 (c1 TIME PRIMARY KEY, c2 INT8);
+INSERT INTO t93396 VALUES ('0:0:0'::TIME, 0);
+
+query error unsupported comparison: time to decimal
+SELECT * FROM t93396 WHERE (c1, c2) = (SELECT 0.0, 0);

--- a/pkg/sql/logictest/tests/local-legacy-schema-changer/generated_test.go
+++ b/pkg/sql/logictest/tests/local-legacy-schema-changer/generated_test.go
@@ -1906,6 +1906,13 @@ func TestLogic_tuple(
 	runLogicTest(t, "tuple")
 }
 
+func TestLogic_tuple_local(
+	t *testing.T,
+) {
+	defer leaktest.AfterTest(t)()
+	runLogicTest(t, "tuple_local")
+}
+
 func TestLogic_txn(
 	t *testing.T,
 ) {

--- a/pkg/sql/logictest/tests/local-vec-off/generated_test.go
+++ b/pkg/sql/logictest/tests/local-vec-off/generated_test.go
@@ -1941,6 +1941,13 @@ func TestLogic_tuple(
 	runLogicTest(t, "tuple")
 }
 
+func TestLogic_tuple_local(
+	t *testing.T,
+) {
+	defer leaktest.AfterTest(t)()
+	runLogicTest(t, "tuple_local")
+}
+
 func TestLogic_txn(
 	t *testing.T,
 ) {

--- a/pkg/sql/logictest/tests/local/generated_test.go
+++ b/pkg/sql/logictest/tests/local/generated_test.go
@@ -2130,6 +2130,13 @@ func TestLogic_tuple(
 	runLogicTest(t, "tuple")
 }
 
+func TestLogic_tuple_local(
+	t *testing.T,
+) {
+	defer leaktest.AfterTest(t)()
+	runLogicTest(t, "tuple_local")
+}
+
 func TestLogic_txn(
 	t *testing.T,
 ) {

--- a/pkg/sql/rowenc/encoded_datum.go
+++ b/pkg/sql/rowenc/encoded_datum.go
@@ -372,7 +372,7 @@ func (ed *EncDatum) Compare(
 	if err := rhs.EnsureDecoded(typ, a); err != nil {
 		return 0, err
 	}
-	return ed.Datum.Compare(evalCtx, rhs.Datum), nil
+	return ed.Datum.CompareError(evalCtx, rhs.Datum)
 }
 
 // GetInt decodes an EncDatum that is known to be of integer type and returns
@@ -541,7 +541,10 @@ func (r EncDatumRow) CompareToDatums(
 		if err := r[c.ColIdx].EnsureDecoded(types[c.ColIdx], a); err != nil {
 			return 0, err
 		}
-		cmp := r[c.ColIdx].Datum.Compare(evalCtx, rhs[c.ColIdx])
+		cmp, err := r[c.ColIdx].Datum.CompareError(evalCtx, rhs[c.ColIdx])
+		if err != nil {
+			return 0, err
+		}
 		if cmp != 0 {
 			if c.Direction == encoding.Descending {
 				cmp = -cmp

--- a/pkg/sql/rowexec/windower.go
+++ b/pkg/sql/rowexec/windower.go
@@ -792,7 +792,10 @@ func (n *partitionPeerGrouper) InSameGroup(i, j int) (bool, error) {
 			n.err = err
 			return false, n.err
 		}
-		if c := da.Compare(n.evalCtx, db); c != 0 {
+		if c, err := da.CompareError(n.evalCtx, db); err != nil {
+			n.err = err
+			return false, n.err
+		} else if c != 0 {
 			if o.Direction != execinfrapb.Ordering_Column_ASC {
 				return false, nil
 			}

--- a/pkg/sql/sem/builtins/aggregate_builtins.go
+++ b/pkg/sql/sem/builtins/aggregate_builtins.go
@@ -3046,7 +3046,10 @@ func (a *maxAggregate) Add(ctx context.Context, datum tree.Datum, _ ...tree.Datu
 		a.max = datum
 		return nil
 	}
-	c := a.max.Compare(a.evalCtx, datum)
+	c, err := a.max.CompareError(a.evalCtx, datum)
+	if err != nil {
+		return err
+	}
 	if c < 0 {
 		a.max = datum
 		if a.variableDatumSize {
@@ -3116,7 +3119,10 @@ func (a *minAggregate) Add(ctx context.Context, datum tree.Datum, _ ...tree.Datu
 		a.min = datum
 		return nil
 	}
-	c := a.min.Compare(a.evalCtx, datum)
+	c, err := a.min.CompareError(a.evalCtx, datum)
+	if err != nil {
+		return err
+	}
 	if c > 0 {
 		a.min = datum
 		if a.variableDatumSize {

--- a/pkg/sql/sem/builtins/math_builtins.go
+++ b/pkg/sql/sem/builtins/math_builtins.go
@@ -611,7 +611,9 @@ var mathBuiltins = map[string]builtinDefinition{
 				}
 
 				for i, v := range thresholds.Array {
-					if operand.Compare(evalCtx, v) < 0 {
+					if cmp, err := operand.CompareError(evalCtx, v); err != nil {
+						return tree.NewDInt(0), err
+					} else if cmp < 0 {
 						return tree.NewDInt(tree.DInt(i)), nil
 					}
 				}

--- a/pkg/sql/sem/builtins/pg_builtins.go
+++ b/pkg/sql/sem/builtins/pg_builtins.go
@@ -566,7 +566,9 @@ var pgBuiltins = map[string]builtinDefinition{
 			},
 			ReturnType: tree.FixedReturnType(types.String),
 			Fn: func(ctx context.Context, evalCtx *eval.Context, args tree.Datums) (tree.Datum, error) {
-				if args[0].Compare(evalCtx, DatEncodingUTFId) == 0 {
+				if cmp, err := args[0].CompareError(evalCtx, DatEncodingUTFId); err != nil {
+					return tree.DNull, err
+				} else if cmp == 0 {
 					return datEncodingUTF8ShortName, nil
 				}
 				return tree.DNull, nil

--- a/pkg/sql/sem/eval/binary_op.go
+++ b/pkg/sql/sem/eval/binary_op.go
@@ -198,7 +198,11 @@ func (e *evaluator) EvalCompareTupleOp(
 				return tree.DNull, nil
 			}
 		} else {
-			cmp = leftElem.Compare(e.ctx(), rightElem)
+			var err error
+			cmp, err = leftElem.CompareError(e.ctx(), rightElem)
+			if err != nil {
+				return tree.DNull, err
+			}
 			if cmp != 0 {
 				break
 			}
@@ -502,7 +506,9 @@ func (e *evaluator) EvalInTupleOp(
 		for _, val := range vtuple.D {
 			if val == tree.DNull {
 				sawNull = true
-			} else if val.Compare(e.ctx(), arg) == 0 {
+			} else if cmp, err := val.CompareError(e.ctx(), arg); err != nil {
+				return tree.DNull, err
+			} else if cmp == 0 {
 				return tree.DBoolTrue, nil
 			}
 		}
@@ -515,10 +521,12 @@ func (e *evaluator) EvalInTupleOp(
 				sawNull = true
 			} else {
 				// Use the EQ function which properly handles NULLs.
-				if res := cmpOpTupleFn(
+				if res, err := cmpOpTupleFn(
 					e.ctx(), *argTuple, *val.(*tree.DTuple),
 					treecmp.MakeComparisonOperator(treecmp.EQ),
-				); res == tree.DNull {
+				); err != nil {
+					return tree.DNull, err
+				} else if res == tree.DNull {
 					sawNull = true
 				} else if res == tree.DBoolTrue {
 					return tree.DBoolTrue, nil

--- a/pkg/sql/sem/eval/comparison.go
+++ b/pkg/sql/sem/eval/comparison.go
@@ -137,7 +137,7 @@ func boolFromCmp(cmp int, op treecmp.ComparisonOperator) *tree.DBool {
 
 func cmpOpTupleFn(
 	ctx tree.CompareContext, left, right tree.DTuple, op treecmp.ComparisonOperator,
-) tree.Datum {
+) (tree.Datum, error) {
 	cmp := 0
 	sawNull := false
 	for i, leftElem := range left.D {
@@ -157,7 +157,7 @@ func cmpOpTupleFn(
 			case treecmp.IsNotDistinctFrom:
 				// For IS NOT DISTINCT FROM, NULLs are "equal".
 				if leftElem != tree.DNull || rightElem != tree.DNull {
-					return tree.DBoolFalse
+					return tree.DBoolFalse, nil
 				}
 
 			default:
@@ -166,10 +166,14 @@ func cmpOpTupleFn(
 				// NULL. This is because NULL is thought of as "unknown" and tuple
 				// inequality is defined lexicographically, so once a NULL comparison is
 				// seen, the result of the entire tuple comparison is unknown.
-				return tree.DNull
+				return tree.DNull, nil
 			}
 		} else {
-			cmp = leftElem.Compare(ctx, rightElem)
+			var err error
+			cmp, err = leftElem.CompareError(ctx, rightElem)
+			if err != nil {
+				return tree.DNull, err
+			}
 			if cmp != 0 {
 				break
 			}
@@ -180,7 +184,7 @@ func cmpOpTupleFn(
 		// The op is EQ and all non-NULL elements are equal, but we saw at least
 		// one NULL element. Since NULL comparisons are treated as unknown, the
 		// result of the comparison becomes unknown (NULL).
-		return tree.DNull
+		return tree.DNull, nil
 	}
-	return b
+	return b, nil
 }

--- a/pkg/sql/sem/eval/window_funcs.go
+++ b/pkg/sql/sem/eval/window_funcs.go
@@ -698,5 +698,5 @@ func compareForWindow(evalCtx *Context, left, right tree.Datum) (int, error) {
 			return 0, err
 		}
 	}
-	return left.Compare(evalCtx, right), nil
+	return left.CompareError(evalCtx, right)
 }

--- a/pkg/sql/sem/tree/eval.go
+++ b/pkg/sql/sem/tree/eval.go
@@ -353,7 +353,9 @@ func ArrayContains(ctx CompareContext, haystack *DArray, needles *DArray) (*DBoo
 		}
 		var found bool
 		for _, hay := range haystack.Array {
-			if needle.Compare(ctx, hay) == 0 {
+			if cmp, err := needle.CompareError(ctx, hay); err != nil {
+				return DBoolFalse, err
+			} else if cmp == 0 {
 				found = true
 				break
 			}
@@ -377,7 +379,9 @@ func ArrayOverlaps(ctx CompareContext, array, other *DArray) (*DBool, error) {
 			continue
 		}
 		for _, hay := range other.Array {
-			if needle.Compare(ctx, hay) == 0 {
+			if cmp, err := needle.CompareError(ctx, hay); err != nil {
+				return DBoolFalse, err
+			} else if cmp == 0 {
 				return DBoolTrue, nil
 			}
 		}

--- a/pkg/sql/stats/bounds/extremes.go
+++ b/pkg/sql/stats/bounds/extremes.go
@@ -92,7 +92,9 @@ func GetUsingExtremesBounds(
 	// but if none exist, return error
 	for i := range histogram {
 		hist := &histogram[i]
-		if hist.UpperBound.Compare(evalCtx, tree.DNull) != 0 {
+		if cmp, err := hist.UpperBound.CompareError(evalCtx, tree.DNull); err != nil {
+			return lowerBound, nil, err
+		} else if cmp != 0 {
 			lowerBound = hist.UpperBound
 			break
 		}

--- a/pkg/sql/stats/histogram.go
+++ b/pkg/sql/stats/histogram.go
@@ -225,7 +225,9 @@ func equiDepthHistogramWithoutAdjustment(
 		// numLess is the number of samples less than upper (in this bucket).
 		numLess := 0
 		for ; numLess < numSamplesInBucket-1; numLess++ {
-			if c := samples[i+numLess].Compare(compareCtx, upper); c == 0 {
+			if c, err := samples[i+numLess].CompareError(compareCtx, upper); err != nil {
+				return histogram{}, err
+			} else if c == 0 {
 				break
 			} else if c > 0 {
 				return histogram{}, errors.AssertionFailedf("%+v", "samples not sorted")
@@ -233,7 +235,9 @@ func equiDepthHistogramWithoutAdjustment(
 		}
 		// Advance the boundary of the bucket to cover all samples equal to upper.
 		for ; i+numSamplesInBucket < numSamples; numSamplesInBucket++ {
-			if samples[i+numSamplesInBucket].Compare(compareCtx, upper) != 0 {
+			if c, err := samples[i+numSamplesInBucket].CompareError(compareCtx, upper); err != nil {
+				return histogram{}, err
+			} else if c != 0 {
 				break
 			}
 		}

